### PR TITLE
Add FedSMOO baseline

### PR DESCRIPTION
**Pull request by kumar-devesh**
### Description

Add FedSMOO baseline to flower.

### Related issues/PRs

Resolves Issue #2017 
 
### Current Status 

Experiments reproduced for FedAvg baseline result in ~10 percent less accuracies than the reported results in the paper

### CODE REVIEW

File: baselines/FedSMOO/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    "Topic :: Software Development :: Libraries",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+    "Typing :: Typed",
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <3.12.0"  # don't change this
+flwr = "1.3.0" # don't change this
+ray = "1.11.1" # don't change this
+hydra-core = "1.3.2" # don't change this
```
Comment (jafermarq) [inline baselines/FedSMOO/pyproject.toml:43]:
```suggestion
python = ">=3.10.0, <3.11.0" 
flwr = { extras = ["simulation"], version = "1.5.0" }
hydra-core = "1.3.2" # don't change this
torch = { url = "https://download.pytorch.org/whl/cu116/torch-1.13.1%2Bcu116-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu116/torchvision-0.14.1%2Bcu116-cp310-cp310-linux_x86_64.whl"}
```

Comment (jafermarq) [reply baselines/FedSMOO/pyproject.toml:43]: Just updating to the latest version of Flower. Also, i'm adding here the torch/torchvision packages. In this way everything is installed via `poetry shell`.

File: baselines/FedSMOO/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,135 @@
+[build-system]
+requires = ["poetry-core>=1.4.0"]
+build-backend = "poetry.masonry.api"
+
+[tool.poetry]
+name = "FedSMOO" # <----- Ensure it matches the name of your baseline directory containing all the source code
```
Comment (jafermarq) [inline baselines/FedSMOO/pyproject.toml:6]:
```suggestion
name = "fedsmoo" # <----- Ensure it matches the name of your baseline directory containing all the source code
```

Comment (jafermarq) [reply baselines/FedSMOO/pyproject.toml:6]: As mentioned in my main comment, please rename your baselines directories to `fedsmoo` (lowercase)

File: baselines/FedSMOO/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+python = ">=3.8.15, <3.12.0"  # don't change this
+flwr = "1.3.0" # don't change this
+ray = "1.11.1" # don't change this
+hydra-core = "1.3.2" # don't change this
+
+[tool.poetry.dev-dependencies]
+isort = "==5.11.5"
+black = "==23.1.0"
+docformatter = "==1.5.1"
+mypy = "==0.961"
```
Comment (jafermarq) [inline baselines/FedSMOO/pyproject.toml:49]:
```suggestion
mypy = "==1.4.1"
```

File: baselines/FedSMOO/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+:warning: _The Python environment for all baselines should follow these guidelines in the `EXTENDED_README`. Specify the steps to create and activate your environment. If there are any external system-wide requirements, please include instructions for them too. These instructions should be comprehensive enough so anyone can run them (if non standard, describe them step-by-step)._
+
+To construct the Python environment follow these steps:
+
+```bash
+# install the base Poetry environment
+poetry install
+
+# activate the environment
+poetry shell
```
Comment (jafermarq) [inline baselines/FedSMOO/README.md:58]:
```suggestion
# Set python version
pyenv install 3.10.6
pyenv local 3.10.6
# Tell poetry to use python 3.10
poetry env use 3.10.6
# install the base Poetry environment
poetry install
# activate the environment
poetry shell
```

Comment (jafermarq) [reply baselines/FedSMOO/README.md:58]: Now baselines should use python3.10+. I thought of providing a bit more intermediate steps showing how to do this with `pyenv` (just as described in the `EXTENDED_README.md`)

File: baselines/FedSMOO/README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+# install PyTorch with GPU support. Please note this baseline is very lightweight so it can run fine on a CPU.
+pip install torch==1.13.1+cu116 torchvision==0.14.1+cu116 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu116
```
Comment (jafermarq) [inline baselines/FedSMOO/README.md:61]:
```suggestion
```

Comment (jafermarq) [reply baselines/FedSMOO/README.md:61]: No longer needed since it is in `pyproject.toml` (see suggestion for that file).

File: baselines/FedSMOO/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# torch==1.13.1+cu116 torchvision==0.14.1+cu116 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu116
```
Comment (jafermarq) [inline baselines/FedSMOO/requirements.txt:1]: Is this `requirements.txt` needed?

File: baselines/FedSMOO/script.sh [hunk 7]
```diff
@@ -0,0 +1,2 @@
+export HYDRA_CONFIG_PATH=conf/fedavg.yaml
```
Comment (jafermarq) [inline baselines/FedSMOO/script.sh:1]: this is not needed really since you use `--config-name=` below

File: baselines/FedSMOO/script.sh [hunk 8]
```diff
@@ -0,0 +1,2 @@
...
+HYDRA_FULL_ERROR=1 python -m FedSMOO.main --config-name=fedavg ++num_rounds=800
```
Comment (jafermarq) [inline baselines/FedSMOO/script.sh:2]:
```suggestion
HYDRA_FULL_ERROR=1 python -m fedsmoo.main --config-name=fedavg num_rounds=800
```

Comment (jafermarq) [reply baselines/FedSMOO/script.sh:2]: just highlighting to update this once your rename everything from `FedSMOO` to `fedsmoo`. Also, `num_rounds` doesn't need the `++`. Or am I missing something?

File: baselines/FedSMOO/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    "Topic :: Software Development :: Libraries :: Python Modules",
+    "Typing :: Typed",
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <3.12.0"  # don't change this
+flwr = "1.3.0" # don't change this
+ray = "1.11.1" # don't change this
+hydra-core = "1.3.2" # don't change this
+
```
Comment (jafermarq) [inline baselines/FedSMOO/pyproject.toml:44]:
```suggestion
tqdm = "4.66.1"
matplotlib = "3.8.0"

```

Comment (jafermarq) [reply baselines/FedSMOO/pyproject.toml:44]: adding missing packages

File: baselines/FedSMOO/FedSMOO/conf/fedavg.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+client:
+  # client config
+  local_epochs: 5
+  learning_rate: 0.1  # local learning rate
+  lr_decay: 0.998     # 0.9995 in demo code
+  stragglers: 0.9
+
+client_resources:
+  num_cpus: 10
+  num_gpus: 1
```
Comment (jafermarq) [inline baselines/FedSMOO/FedSMOO/conf/fedavg.yaml:47]:
```suggestion
  num_cpus: 4
  num_gpus: 0.2
```

Comment (jafermarq) [reply baselines/FedSMOO/FedSMOO/conf/fedavg.yaml:47]: When running your code i noticed clients just use 1GB of VRAM. I think it's fine if we allow for several clients to run in parallel on the same GPU. This can be done easily by lowering the fraction of GPU resources needed per client. I propose allowing to pack 5 (so `num_gpus=0.2`, since `1/0.2 = 5`). Also, let's lower CPU resources that each client needs. I don't think more than 4 cpus is really needed in this setting.

You can read more about how to set the `client_resources` effectively in the documentation: https://flower.dev/docs/framework/how-to-run-simulations.html#virtualclientengine-resources.

File: baselines/FedSMOO/FedSMOO/main.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    # also other buffers, statistics, info needed to be saved in order to later
+    # on generate the plots you provide in the README.md. You can for instance
+    # access elements that belong to the strategy for example:
+    # data = strategy.get_my_custom_data() -- assuming you have such method defined.
+    # Hydra will generate for you a directory each time you run the code. You
+    # can retrieve the path to that directory with this:
+    # save_path = HydraConfig.get().runtime.output_dir
+
+    accuracy_test = history.metrics_centralized["accuracy"]
+    plot_fn(cfg, accuracy_test)
```
Comment (jafermarq) [inline baselines/FedSMOO/FedSMOO/main.py:159]: This is currently crashing because `./results` (hardcoded inside that function) doesn't exist. You can easily get the path for the experiment directory that's automatically generated by `Hydra` via:

```python
from hydra.core.hydra_config import HydraConfig
save_path = HydraConfig.get().runtime.output_dir
```

As you have probably noticed, a directory in `outputs/<current-date>/<current-time>` is generated each time you run your `main.py`. the `save_path` above will return you the path to that directory. Use that to save results and plots

File: baselines/FedSMOO/FedSMOO/main.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    else:
+        dataset_prep = "pathological"
+
+    logfile = f"./results/{cfg.method}_{cfg.dataset.dataset_name}_{cfg.num_clients}_{dataset_prep}.log"
+    logging.basicConfig(filename=logfile, 
+					format='%(asctime)s %(message)s', 
+					filemode='w') 
+    logger=logging.getLogger() 
+    logger.setLevel(logging.DEBUG) 
+    logger.info(accuracy_test)
```
Comment (jafermarq) [inline baselines/FedSMOO/FedSMOO/main.py:172]: Hydra creates a log for you. Although the current integration of hydra+flower is not 100% great, do you need the above?

File: baselines/FedSMOO/script.sh (hunk 8 above)
Comment (jafermarq) [reply baselines/FedSMOO/script.sh:2]: Another comment regarding `--config-name` usage here. Since your default `config_name` is already `"fedavg"` in your `main.py`, there's no need make use of the `--config-name` flag. However you'll need it if you want to use your `fedsmoo` config.

File: baselines/FedSMOO/FedSMOO/conf/fedsmoo.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  learning_rate: 0.1 # local learning rate
+  weight_decay: 0.001
+  sch_step: 1
+  sch_gamma: 1.0
+  alpha: 0.1
+  lr_decay: 0.998     # 0.9995 in demo code
+
+client_resources:
+  num_cpus: 4
+  num_gpus: 1
```
Comment (jafermarq) [inline baselines/FedSMOO/FedSMOO/conf/fedsmoo.yaml:50]: Please consider adjusting these based on my comments for the `fedavg.yaml` config. On my tests i see your `fedsmoo` clients using a bit more memory (~1.4GB) but require more compute (to be expected). I'd say setting `num_gpus` to either `0.2` is still safe.

File: baselines/FedSMOO/FedSMOO/utils.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+"""
+
+import matplotlib.pyplot as plt
+from typing import List, Tuple
+
+def plot_fn(cfg, acc: List[Tuple[int, float]]):
+    x_val = [x[0] for x in acc]
+    y_val = [x[1] for x in acc]
+
+    plt.plot(x_val,y_val,'or')
```
Comment (jafermarq) [inline baselines/FedSMOO/FedSMOO/utils.py:15]:
```suggestion
    plt.plot(x_val,y_val,'or')
    plt.plot(x_val,y_val,'-r')
    plt.grid()
```

Comment (jafermarq) [reply baselines/FedSMOO/FedSMOO/utils.py:15]: Probably you are still working on this. I addition to those changes, don't forget to add meaningful labels for both axis and a title.

**Review by jafermarq — COMMENTED** [review]
Hi @kumar-devesh,

I just completed the first pass through your `FedSMOO` implementation. I was able to run the experiments for a few rounds (after doing some small changes -- see my code suggestions below). This is a nice start!

You'll see i've made quite a few suggestions to change parts of the codebase, these are small changes but do add up. Most of them in the `README.md`, `pyproject.toml` and the configs.

In addition to those, please rename your baseline from `FedSMOO` to `fedsmoo`. Probably the most convenient way of doing this is with `git mv`:
```bash
git mv FedSMOO fedsmoo # do this for your top-level directory and the inner one too
```
Then, renaming all your `import FedSMOO.<...>` can be done very easily with a IDE using standard _string_ search-and-replace.

I applied all the changes suggested and i was able to run your code (I only ran it for 10 rounds, for both configs).

Please ping me again when you have made the changes, happy to discuss them further.

**Comment by kumar-devesh** [comment]
updated the code based on the comments in your review

File: baselines/fedsmoo/script.sh
Comment (jafermarq) [inline baselines/fedsmoo/script.sh:1]: Unless the code is really complicated, it is preferable to put this in the `readme`.

File: baselines/fedsmoo/fedsmoo/models.py [hunk 15]
```diff
@@ -0,0 +1,426 @@
...
+
+#         Returns
+#         --------------
+#         torch.Tensor
+#             model output
+#         """
+
+#         x = self.resnet18(input_tensor)
+#         return x
+
```
Comment (jafermarq) [inline baselines/fedsmoo/fedsmoo/models.py:60]: remove if not needed?

**Review by jafermarq — COMMENTED** [review]
Hi @kumar-devesh,

I have added a few extra comments. I found the partitioning function a bit complex and left some suggestions there. It might be worth double checking it in isolation to see if the partitions are created properly (specially if you see your code performing worse than the paper). 
Ping me once you have the readme ready with some results. Happy to discuss about your baseline at any point.

**Comment by jafermarq** [comment]
Hi, the _Summer of Reproducibility_ is now over. As per our last messages in our Slack channel, we are closing a few PRs including yours. We still think that reproducing your baseline has value to the FL community at large so you are welcome to resume your work from were you left it if you want by creating a new PR. Note the associated Issue will remain open. Thanks!
